website/docs/script/type-id.mdx

```diff
@@ -0,0 +1,237 @@
+
+We can use a Script structure! Since lock Script is typically used for signature verification, we can use a type Script for this problem. A type Script can stay perfectly unchanged when a cell's data get changed. Hence we added `hash_type` field to CKB's Script structure, and modified the Script resolving flow to the following:
+
+- For each dep cell, we extract the `test hash` based on `hash_type` value in the Script structure:
+  - If `hash_type` is `data`, the dep cell's data hash is used as `test hash`
+  - If `hash_type` is `type`, the dep cell's type Script hash is used as `test hash`
+- CKB extracts `code_hash` and `hash_type` values from the Script structure to run
+- If CKB finds a dep cell whose `test hash` matches specified `code_hash`, CKB uses the data in the found dep cell as the scrip to run.
+- If no dep cell has `test hash` matching specified `code_hash`, CKB results in a validation error.
+
+Notice the `hash_type` used here, is the value belonging to the Script to run, not the values of scripts in a dep cell. You could perfectly have 2 inputs in a transaction, one using `data` as `hash_type`, the other using `type` as `hash_type`. Either one of them will use its own correct way to locate the correct cell.
+
+This way we have totally conquered determinism property A above, but for property B there will be some subtle affects, we will discuss those in more details below.
+
+## Putting Everything Together
+
+There's still one problem unsolved, there might still be an attack:
+
+- A lock Script L1 is stored in a cell C1 with type Script T1
+- Alice guards her cells via lock Script L1 using `hash_type` as `type`, by definition, she fills her Script structure's `code_hash` field with the hash of type Script T1
+- Bob creates a different cell C2 with a always success lock Script L2, and also use the same type Script T1
+- Now Bob can use C2 as a dep cell to unlock Alice's cell. CKB won't be able to distinguish that Alice wants to use C1, while Bob provides C2. Both C1 and C2 use T1 as type Script
+
+That teaches us a very important lesson: if you build a lock/type Script and want people to leverage the upgradability property, you have to make sure to use a type Script that is unique and unforgeable.
+
+Luckily, we just solved this problem! In the above we just developed a type Script that can provide a unique type Script hash in CKB. Combined together, this unique type Script and `hash_type` feature in CKB provide a way to upgrade already deployed smart contracts.
+
+## Type-ID as a Genesis Script
+
+The Type ID type Script can be implemented as an ordinaly Script, but CKB team choose to implement it in the CKB node as a special genesis Script. Because if we want to make it upgradable, it has to use itself as the type Script via type Script hash, which is a recursive dependency.
+
+As a genesis Script, the Type ID code cell uses a special type Script hash, which is just the ascii codes in hex of the text `TYPE_ID`.
```

Comment:
```suggestion
The Type ID type Script can be implemented as an ordinary Script, but the CKB team chose to implement it in the CKB node as a special genesis Script. This is because if we want to make it upgradable, it has to use itself as the Type Script via Type Script hash, which creates a recursive dependency.

As a genesis Script, the Type ID code Cell uses a special Type Script hash, which is just the ASCII codes in hex of the text `TYPE_ID`.
```